File: java/MajorityNumber3.java
//http://www.lintcode.com/en/problem/majority-number-iii/
import java.util.Iterator;
import java.util.HashMap;
import java.util.ArrayList;
import java.util.Map;

public class SolutionMajorityNumber3 {
    /**
     * @param nums: A list of integers
     * @param k: As described
     * @return: The majority number
     */
    public int majorityNumber(ArrayList<Integer> nums, int k) {
        // write your code
        int len = nums.size();
        if(len < 1 || k < 1)return -1;
        HashMap<Integer, Integer> kth = new HashMap<>();
        for(int i : nums) {
            if(kth.size() < k - 1) { // has at most k - 1 majority numbers that satisfy more than 1 / k
                if(kth.get(i) == null)kth.put(i, 1);
                else kth.put(i, kth.get(i) + 1);
            }
            else {
                if(kth.get(i) != null)kth.put(i, kth.get(i) + 1); 
                else {
                    Iterator<Map.Entry<Integer, Integer> > itor = kth.entrySet().iterator();
                    while(itor.hasNext()) {
                        Map.Entry<Integer, Integer> pair = itor.next();
                        if(pair.getValue() > 1)
                            pair.setValue(pair.getValue() - 1);
                        else
                            itor.remove();//remove pair during loop using Iterator
                    }
                }
            }
        }
        if(kth.size() > 0) {
            Iterator<Map.Entry<Integer, Integer> > itor = kth.entrySet().iterator();
            int s = len / k, cnt = 0;
            while(itor.hasNext()) {
                Map.Entry<Integer, Integer> pair = itor.next();
                cnt = 0;
                for(int i : nums)
                    if(i == pair.getKey())++cnt;
                if(cnt > s)return pair.getKey();
            }
        }
        return -1;
    }
}

File: python/MajorityNumber3.py
#http://www.lintcode.com/en/problem/majority-number-iii/
class SolutionMajorityNumber3:
    """
    @param nums: A list of integers
    @param k: As described
    @return: The majority number
    """
    def majorityNumber(self, nums, k):
        # write your code here
        l = len(nums)
        if l < 1 or k < 1:
            return -1
        kth = {}
        for i in nums:
            if len(kth) < k - 1:
                if i in kth:
                    kth[i] += 1
                else:
                    kth[i] = 1
            else:
                if i in kth:
                    kth[i] += 1
                else:
                    for key in kth.keys():
                        kth[key] -= 1
                        if kth[key] == 0:
                            del kth[key]
        if len(kth) > 0:
            s = l // k
            cnt = 0
            for key in kth.keys():
                for i in nums:
                    if i == key:
                        cnt += 1
                #print "{0},{1}".format(key, cnt)
                if cnt > s:
                    return key
                else:
                    cnt = 0
        return -1
